authority/admin/api/policy.go

```diff
@@ -0,0 +1,396 @@
+	if err := par.auth.UpdateProvisioner(ctx, prov); err != nil {
+		var pe *authority.PolicyError
+		isPolicyError := errors.As(err, &pe)
+		if isPolicyError && pe.Typ == authority.AdminLockOut || pe.Typ == authority.EvaluationFailure || pe.Typ == authority.ConfigurationFailure {
+			render.Error(w, admin.WrapError(admin.ErrorBadRequestType, pe, "error creating provisioner policy"))
+			return
+		}
+
+		render.Error(w, admin.WrapErrorISE(err, "error creating provisioner policy"))
+		return
+	}
+
+	render.ProtoJSONStatus(w, newPolicy, http.StatusCreated)
+}
+
+// UpdateProvisionerPolicy handles the PUT /admin/provisioners/{name}/policy request
+func (par *PolicyAdminResponder) UpdateProvisionerPolicy(w http.ResponseWriter, r *http.Request) {
+
+	ctx := r.Context()
+	prov := linkedca.ProvisionerFromContext(ctx)
+
+	if prov.Policy == nil {
+		render.Error(w, admin.NewError(admin.ErrorNotFoundType, "provisioner policy does not exist"))
+		return
+	}
+
+	var newPolicy = new(linkedca.Policy)
+	if err := read.ProtoJSON(r.Body, newPolicy); err != nil {
+		render.Error(w, err)
+		return
+	}
+
+	newPolicy.Deduplicate()
+
+	prov.Policy = newPolicy
+	if err := par.auth.UpdateProvisioner(ctx, prov); err != nil {
+		var pe *authority.PolicyError
+		isPolicyError := errors.As(err, &pe)
+		if isPolicyError && pe.Typ == authority.AdminLockOut || pe.Typ == authority.EvaluationFailure || pe.Typ == authority.ConfigurationFailure {
+			render.Error(w, admin.WrapError(admin.ErrorBadRequestType, pe, "error updating provisioner policy"))
+			return
+		}
+
+		render.Error(w, admin.WrapErrorISE(err, "error updating provisioner policy"))
+		return
+	}
+
+	render.ProtoJSONStatus(w, newPolicy, http.StatusOK)
+}
+
+// DeleteProvisionerPolicy handles the DELETE /admin/provisioners/{name}/policy request
+func (par *PolicyAdminResponder) DeleteProvisionerPolicy(w http.ResponseWriter, r *http.Request) {
+
+	ctx := r.Context()
+	prov := linkedca.ProvisionerFromContext(ctx)
+
+	if prov.Policy == nil {
+		render.Error(w, admin.NewError(admin.ErrorNotFoundType, "provisioner policy does not exist"))
+		return
+	}
+
+	// remove the policy
+	prov.Policy = nil
+
+	if err := par.auth.UpdateProvisioner(ctx, prov); err != nil {
+		render.Error(w, admin.WrapErrorISE(err, "error deleting provisioner policy"))
+		return
+	}
+
+	render.JSONStatus(w, DeleteResponse{Status: "ok"}, http.StatusOK)
+}
+
+func (par *PolicyAdminResponder) GetACMEAccountPolicy(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	eak := linkedca.ExternalAccountKeyFromContext(ctx)
+
+	policy := eak.GetPolicy()
+	if policy == nil {
+		render.Error(w, admin.NewError(admin.ErrorNotFoundType, "ACME EAK policy does not exist"))
+		return
+	}
+
+	render.ProtoJSONStatus(w, policy, http.StatusOK)
+}
+
+func (par *PolicyAdminResponder) CreateACMEAccountPolicy(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	prov := linkedca.ProvisionerFromContext(ctx)
+	eak := linkedca.ExternalAccountKeyFromContext(ctx)
```

Comment:
I think we should consolidate on this, it's safer:
```golang
func ExternalAccountKeyFromContext(ctx context.Context) (eak *EABKey, ok bool) {
   eak, ok = ctx.Value(externalAccountKeyContextKey).(*EABKey)
   return
}
```

Reply:
In the locations I use these functions we would have to add a check for the `bool`. What error would you return then?

Reply:
In my context changes I'm generally adding this:

```golang
type dbKey struct{}

// NewDatabaseContext adds the given acme database to the context.
func NewDatabaseContext(ctx context.Context, db DB) context.Context {
	return context.WithValue(ctx, dbKey{}, db)
}

// DatabaseFromContext returns the current acme database from the given context.
func DatabaseFromContext(ctx context.Context) (db DB, ok bool) {
	db, ok = ctx.Value(dbKey{}).(DB)
	return
}

// MustDatabaseFromContext returns the current database from the given context.
// It will panic if it's not in the context.
func MustDatabaseFromContext(ctx context.Context) DB {
	if db, ok := DatabaseFromContext(ctx); !ok {
		panic("acme database is not in the context")
	} else {
		return db
	}
}
```


I think we can standardize on something like that.